### DISCUSSION TITLE
Why is the ZeRO3 optimizer only capable of certain types? Any special reason?

### DISCUSSION THREAD

**Discussion by yhna940** [hpcaitech/ColossalAI#3449] [Community | Q&A]
Why is the ZeRO3 optimizer only capable of certain types? Any special reason?

https://github.com/hpcaitech/ColossalAI/blob/ffcdbf0f6519366f322f2809e19f741492779c6c/colossalai/zero/gemini/gemini_optimizer.py#L72

What can I do if I want to use this optimizer more generically? (I also want to use optimizers like RMS or SGD.)

**Comment by 1SAA** [comment]
Hi, @yhna940 
The reason is that the gradient generated is FP16, and the master weight is FP32. We just need a optimizer that can accept FP16 gradients with FP32 parameters.

**Reply by yhna940** [reply]
@1SAA 
Thank you for answer.
Can't we just cast the gradient to the master parameter's type, like low level zero?
I would be grateful if you could provide feedback on my proposed changes.
https://github.com/hpcaitech/ColossalAI/pull/3699
